### ISSUE TITLE
(Master Regression) ASI HPI Issue

### ISSUE THREAD

**Issue by deltecent** [ElvishArtisan/rivendell#313]
I've been having a problem with my ASI cards under 3.0alpha. Here's the log:
```
Nov 22 00:37:22 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x102.
Nov 22 00:37:22 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x102.
Nov 22 00:37:22 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10c.
Nov 22 00:37:22 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10d.
Nov 22 00:37:30 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x106.
Nov 22 00:37:32 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x102.
Nov 22 00:37:32 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x102.
Nov 22 00:37:32 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10c.
Nov 22 00:37:32 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10d.
Nov 22 00:37:40 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x106.
Nov 22 00:38:26 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x108.
Nov 22 00:38:26 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x105.
Nov 22 00:38:28 rdhost kernel: snd_asihpi 0000:03:00.0: Adapter at index 0 is not in low latency mode
Nov 22 00:38:28 rdhost kernel: snd_asihpi 0000:03:00.0: probe succeeded for ASI5611 HPI index 0
Nov 22 00:38:28 rdhost kernel: snd_asihpi 0000:03:00.0: cannot find the slot for index 0 (range 0-0), error: -16
Nov 22 00:38:28 rdhost kernel: snd_asihpi 0000:03:00.0: Adapter index 0->ALSA index 1
Nov 22 00:38:39 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x102.
Nov 22 00:38:39 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x102.
Nov 22 00:38:39 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10c.
Nov 22 00:38:39 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10d.
```
Here's ```dmsg``` which has some interesting stuff in it:
```
[    5.400880] ASIHPI driver 4.14.03
[    5.400885] hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x108.
[    5.400934] snd_asihpi 0000:03:00.0: probe 104c:a106,175c:5600,0000
[    5.407715] hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x105.
[    5.422757] snd_asihpi 0000:03:00.0: Firmware version mismatch: DSP image 40D00 != Driver 40E03
[    5.484557] snd_asihpi 0000:03:00.0: Firmware version mismatch: DSP image 40D00 != Driver 40E03
[    7.027889] snd_asihpi 0000:03:00.0: Adapter at index 0 is not in low latency mode
[    7.027955] snd_asihpi 0000:03:00.0: using polled mode
[    7.027958] snd_asihpi 0000:03:00.0: probe succeeded for ASI5611 HPI index 0
[    7.027970] snd_asihpi 0000:03:00.0: cannot find the slot for index 0 (range 0-0), error: -16
[    7.028059] snd_asihpi 0000:03:00.0: Adapter index 0->ALSA index 1
[    7.028647] snd_asihpi 0000:03:00.0: Has dma:1, grouping:1, mrx:0, uif:128
[    7.033558] snd_asihpi 0000:03:00.0: 49 mixer controls found
[    7.071762] WARNING: module 'asihpi' built without retpoline-enabled compiler, may affect Spectre v2 mitigation
[    7.072698] asihpi: applying kernel_stack fix up
[    7.074309] ASIHPI driver 4.13.00
[   18.580600] hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x102.
[   18.580617] hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x102.
[   18.580626] hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10c.
[   18.580631] hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10d.
```

**Comment by deltecent** [comment]
I'm thinking these errors may be benign. What is happening is audio is randomly going away in RDAirplay and I thought these errors may be related. Here's an example:

RDAirplay starts playing ```/var/snd/010456_001.wav```:
```
Nov 22 01:40:36 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10c.
Nov 22 01:40:36 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10d.
Nov 22 01:40:36 rdhost caed: LoadPlayback  Card: 0  Stream: 1  Name: /var/snd/010456_001.wav  Handle: 27
Nov 22 01:40:36 rdhost rdairplay: started audio cart: Line: 10  Cart: 10456  Cut: 1 Pos: 0  Card: 0  Stream: 1  Port: 0
Nov 22 01:40:36 rdhost caed: PlaybackPosition - Card: 0  Stream: 1  Pos: 0  Handle: 27
Nov 22 01:40:36 rdhost caed: Play - 1
Nov 22 01:40:36 rdhost caed: Play - Card: 0  Stream: 1  Handle: 27  Length: 225098  Speed: 100000  Pitch: 0
```
Then I start playing ```/var/snd/010001_001.wav``` in RDLibrary:
```
Nov 22 01:41:06 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10c.
Nov 22 01:41:06 rdhost kernel: hpimsgx.c:169 suspicious adapter index 0 in subsys message 0x10d.
Nov 22 01:41:06 rdhost caed: LoadPlayback  Card: 0  Stream: 0  Name: /var/snd/010001_001.wav  Handle: 28
Nov 22 01:41:06 rdhost caed: PlaybackPosition - Card: 0  Stream: 0  Pos: 3288  Handle: 28
Nov 22 01:41:06 rdhost caed: Play - 1
Nov 22 01:41:06 rdhost caed: Play - Card: 0  Stream: 0  Handle: 28  Length: 215042  Speed: 100000  Pitch: 0
```
Randomly when doing this, the audio from RDAirplay goes away. RDAirplay shows the audio is still playing, but it is muted or something. As soon as RDAirplay finishes with that event, the logs look normal, and the next event fires and can be heard:
```
Nov 22 01:44:21 rdhost caed: StopPlayback - Card: 0  Stream: 1  Handle: 27
Nov 22 01:44:21 rdhost rdairplay: finished event: Line: 10  Cart: 10456  Cut: 1 Card: 0  Stream: 1  Port: 0
Nov 22 01:44:21 rdhost caed: StopPlayback - Card: 0  Stream: 1  Handle: 27
Nov 22 01:44:21 rdhost caed: UnloadPlayback - Card: 0  Stream: 1  Handle: 27
```

**Comment by deltecent** [comment]
This is what appears to be going on.

When speaking with the HPI driver, there is an ```hpi_stream``` variable that is returned from ```HPI_OutStreamOpen```.

In my case, when I start playing audio, I get a value of 12288. If I start another stream by playing something in RDLibrary, it gets a value to 12289 and everything works fine, both streams continue to play.

If the only thing playing is 12289 and then another stream starts with a value of 12288, the 12288 stream plays and the 12289 stream stops playing, as if all streams above the stream that starts are stopping.

I have no idea why. At least I can now reproduce it at will.

**Comment by deltecent** [comment]
This will probably need a second set of eyes:

rd.h defines RD_MAX_PORTS as 24.
asihpi/hpi.h defines HPI_MAX_NODES as 8.

First Question: What's the difference between a "port" and a "node"?

Here's the problem, using RDSimplePlayer as an example:

```
      for(int i=0;i<RD_MAX_PORTS;i++) {
        play_cae->setOutputVolume(play_card,play_stream,i,RD_MUTE_DEPTH);
      }
```
This is going to call setOutputVolume on "ports" 0-23.

Now here's what's in rdhpisoundcard.h:
```
  bool output_stream_volume[HPI_MAX_ADAPTERS][HPI_MAX_STREAMS][HPI_MAX_NODES];
```
Each adapter has an array of 8 nodes (ports).

Now here's what's going on in rdhpisoundcard.cpp:
```
void RDHPISoundCard::setOutputVolume(int card,int stream,int port,int level)
{
  if(!haveOutputVolume(card,stream,port)) {
    return;
  }
  short gain[2];
  gain[0]=level;
  gain[1]=level;
  LogHpi(HPI_VolumeSetGain(NULL,
                           output_stream_volume_control[card][stream][port],
                           gain),__LINE__);
}
```
This is going to pass through the requested stream and the next 2 streams (because once it passes port 7, it's going to overrun the 8 nodes) setting the volume, which is why when the first stream is playing and the second stream starts, everything is fine. When the second stream is playing and the first stream starts, it kills the volume of the new stream and the following 2 streams.

I fixed the problem by changing ```RDHPISoundCard::setOutputVolume```:
```
if(!haveOutputVolume(card,stream,port)**||port>=HPI_MAX_NODES**) {
```

The real question: should RD_MAX_PORTS match HPI_MAX_NODES?

**Comment by deltecent** [comment]
Fixed #320

**Comment by deltecent** [comment]
Fixed in PR #320
